Client/Frontend/Home/ActivityStreamPanel.swift

```diff
@@ -0,0 +1,354 @@
+        tableView.sectionHeaderHeight = UITableViewAutomaticDimension
+        return tableView
+    }()
+
+    private let topSiteHandler = ASHorizontalScrollSource()
+
+    //once things get fleshed out we can refactor and find a better home for these
+    var topSites: [TopSiteItem] = []
+    var history: [Site] = []
+
+
+    init(profile: Profile) {
+        self.profile = profile
+        super.init(nibName: nil, bundle: nil)
+
+        self.profile.history.setTopSitesCacheSize(Int32(ASPanelUX.topSitesCacheSize))
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(TopSitesPanel.notificationReceived(_:)), name: NotificationFirefoxAccountChanged, object: nil)
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(TopSitesPanel.notificationReceived(_:)), name: NotificationProfileDidFinishSyncing, object: nil)
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(TopSitesPanel.notificationReceived(_:)), name: NotificationPrivateDataClearedHistory, object: nil)
+        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(TopSitesPanel.notificationReceived(_:)), name: NotificationDynamicFontChanged, object: nil)
+    }
+
+    deinit {
+        NSNotificationCenter.defaultCenter().removeObserver(self, name: NotificationFirefoxAccountChanged, object: nil)
+        NSNotificationCenter.defaultCenter().removeObserver(self, name: NotificationProfileDidFinishSyncing, object: nil)
+        NSNotificationCenter.defaultCenter().removeObserver(self, name: NotificationPrivateDataClearedHistory, object: nil)
+        NSNotificationCenter.defaultCenter().removeObserver(self, name: NotificationDynamicFontChanged, object: nil)
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        reloadTopSites()
+
+        reloadRecentHistory()
+
+        view.addSubview(tableView)
+        tableView.snp_makeConstraints { (make) in
+            make.edges.equalTo(self.view)
+        }
+    }
+
+    override func traitCollectionDidChange(previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        self.topSiteHandler.currentTraits = self.traitCollection
+    }
+
+}
+
+// MARK: -  Section management
+extension ActivityStreamPanel {
+
+    enum Section: Int {
+        case TopSites
+        case History
+
+        static let count = 2
+
+        var title: String? {
+            switch self {
+            case .History: return "Recent Activity"
+            case .TopSites: return nil
+            }
+        }
+
+        var headerHeight: CGFloat {
+            switch self {
+            case .History: return 40
+            case .TopSites: return 0
+            }
+        }
+
+        func cellHeight(traits: UITraitCollection) -> CGFloat {
```

Comment:
Nice! This looks like it's much easier to work with instead of the loop we were doing before with reloading the table view.